compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -934,6 +957,58 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
             .collect();
         let mut fields = fields?;
 
+        if let AggregateTy::Def(_, _) = &mut ty
+            && !fields.is_empty()
+        {
+            // All fields must correspond one-to-one and come from the same aggregate value.
+            if let Value::Projection(copy_from_value, _) = *self.get(fields[0])
+                && fields.iter().enumerate().all(|(index, &v)| {
+                    if let Value::Projection(pointer, ProjectionElem::Field(from_index, _)) =
+                        *self.get(v)
+                        && copy_from_value == pointer
+                        && from_index.index() == index
+                    {
+                        return true;
+                    }
+                    false
+                })
+            {
+                let mut projection = SmallVec::<[PlaceElem<'tcx>; 1]>::new();
+                let mut copy_from_local_value = copy_from_value;
+                loop {
+                    if let Some(local) = self.try_as_local(copy_from_local_value, location) {
+                        projection.reverse();
+                        if let Some(ProjectionElem::Downcast(_, _)) = projection.last() {
+                            projection.pop();
+                        }
```

Comment:
Can the check for downcast be hoisted just before the loop?

Reply:
Done, I created a large if statement.